### PR TITLE
Update collectd and graphite UDP listeners with perf enhancements

### LINKED ISSUE #4678: Add SetReadBuffer() to collectd and Graphite
Started with https://github.com/influxdb/influxdb/pull/4676/

@sparrc discovered we can increase the OS buffers to improve performance.

### PR DESCRIPTION
**Pull request by sparrc**
closes #4678 

related to #4676 

cc @otoolep

### CODE REVIEW

File: services/collectd/service.go [hunk 1]
```diff
@@ -53,7 +53,7 @@ type Service struct {
 	wg      sync.WaitGroup
 	err     chan error
 	stop    chan struct{}
-	ln      *net.UDPConn
+	conn    *net.UDPConn
```
Comment (sparrc) [inline services/collectd/service.go]: changed this just for consistency with graphite and udp services

**Comment by sparrc** [comment]
cc @dswarbrick @daviesalex @nkatsaros

**Comment by otoolep** [comment]
We need the CHANGELOG updated, and https://github.com/influxdb/influxdb/blob/master/etc/config.sample.toml updated to include these new values.

**Comment by otoolep** [comment]
CHANGELOG should be updated in the "Features" section -- allow read buffer size to be set.

File: services/collectd/config.go [hunk 2]
```diff
@@ -7,19 +7,29 @@ import (
...
+	// DefaultBatchPending is the default number of pending write batches.
+	DefaultBatchPending = 10
 
+	// DefaultBatchTimeout is the default batch timeout.
 	DefaultBatchDuration = toml.Duration(10 * time.Second)
 
 	DefaultTypesDB = "/usr/share/collectd/types.db"
+
+	// DefaultUDPReadBuffer is the default UDP read buffer
+	// increasing this increases the number of UDP packets that can be handled.
```
Comment (otoolep) [inline services/collectd/config.go]: This is a little vague. Why not be precise and repeat the text in the Go docs?

"Sets the size of the operating system's receive buffer associated with the UDP traffic "

**Comment by otoolep** [comment]
Does the buffer size apply to TCP traffic? If so, we should set it for Graphite TCP, right?

File: services/collectd/config.go (hunk 2 above)
Comment (sparrc) [inline services/collectd/config.go]: kk :+1:

**Comment by sparrc** [comment]
For graphite we can't change the tcp read buffer at the moment. We would have the change the `net.Listen` call to a `net.DialTCP` call to get that functionality.

**Comment by dswarbrick** [comment]
Somewhere deep down, SetReadBuffer() will be doing the equivalent of a setsockopt() with SO_RCVBUF option. Note that Linux (presumably other *nixes) has maximums that this can be set to. The following are from my workstation, running kernel 4.2:
```
net.core.rmem_default = 212992
net.core.rmem_max = 212992
net.ipv4.tcp_rmem = 4096	87380	6291456
```

**Comment by nkatsaros** [comment]
You can do `tcpConn, ok := conn.(*net.TCPConn)` and do whatever you want on the connection that `s.ln.Accept()` spits out. There's no need to stray from the generic Listen/Accept/Dial functions in the net package.

**Comment by sparrc** [comment]
@dswarbrick That is also what I get:

```
$ sudo sysctl -a | grep net.core.rmem
net.core.rmem_default = 212992
net.core.rmem_max = 212992
```

maybe it can adjust the max? I'm not sure...

**Comment by sparrc** [comment]
@nkatsaros, from what I understand, `net.Listen("tcp"...)` is a simplified alternative to running `DialTCP(...)`. I was just saying that it would be easier to adjust this if we used `DialTCP`, I suppose we could typecast and adjust on every for-loop iteration but that seems messy.

**Comment by dswarbrick** [comment]
@sparrc Need to be root to modify the OS defaults / max. These should not be fiddled with directly by InfluxDB, as they will affect all future sockets, in all applications. It would be worth documenting how to change them, or linking to relevant documentation.

**Comment by sparrc** [comment]
I'll need to test on Linux, TBH I don't know how Go adjusts the read buffer size. It worked on my mac workstation which means it probably works on any BSD system.

**Comment by sparrc** [comment]
oh, looks like BSD already has the max set higher, that would explain why it worked

OSX
```
$ sysctl kern.ipc.maxsockbuf
kern.ipc.maxsockbuf: 8388608
```

FreeBSD:
```
$ sysctl kern.ipc.maxsockbuf
kern.ipc.maxsockbuf: 2097152
```

Get it together Linux!!

I'll put a note in the documentation

**Comment by otoolep** [comment]
Thanks for the heads-up @dswarbrick 

@sparrc -- looks like we have documentation issue here.

File: services/udp/README.md [hunk 3]
```diff
@@ -1,13 +1,125 @@
...
+
+Changes to /etc/sysctl.conf do not take effect until reboot.  To update the values immediately, type the following commands as root:
+
+```
+sysctl -w net.core.rmem_max=8388608
+```
+
+### BSD/Darwin
+
+On BSD/Darwin systems you need to add about a 15% padding to the kernel limit
```
Comment (sparrc) [inline services/udp/README.md]: @otoolep finally tracked down the BSD/Darwin failures I was seeing to this. This PR should be good to merge now.

**Pull request merged**